Add compatibility for Pan-ukbb external data, make column names configurable.

Lipastomies opened the pull request:
This PR aims to make newly matched pan-ukbb data usable in the pheweb browser.

## Changes
- Make external data column names configurable
    - For matrix data, pval and beta are configurable, single files have c:p:r:a + pval and beta
- Add both pval and beta to phenoview tables
    - Beta should now be usable as pan-ukbb scans were made with saige
- In case only matrix DAO or file DAO is available, use that for both operations, otherwise use both (previously matrix DAO was saved to jeeves.ukbb_matrix_dao and file DAO was saved to jeeves.ukbb_dao, so you had to have both for full ukbb functionality. Now they are saved similarly, but if one is missing then either ukbb_matrix_dao or okbb_dao is pointed to be the existing DAO.)
- Example config for pan-ukbb data in config.r5.py
## Possible changes to be made before merge
- Due to the fact that our data has changed quite a lot since R1 when the previous ukbb data was created (remove chr from vars, change to numerical chromosomes from XYMT), the data differ by quite a bit. I had to add a few variables and checks to the DAO initialization & configs. We should see whether this could be done better.
    - file DAO config has `chrom_is_numeric` flag for whether chromosomes are XYMT or 23,24,25
    - matrix DAO has `column_names` dict for pval and beta names, file DAO has c:p:r:a also
    - Whether chrom has "chr" prefix is figured out at runtime 
-  Config file format could be changed
- Frontend table can be prettier

R6 data with pan-ukbb data can be seen in dev.finngen.fi

Fedja commented:
@Lipastomies  add OR also to gene page main table and coding variant table

Fedja commented on pheweb/serve/data_access/db.py:
```diff
@@ -886,9 +896,11 @@ def __init__(self, manifest):
                     self.results[ line[name_idx] ] = ExternalFileResultDao.ResRecord( line[name_idx], line[ncase_idx], line[ncontrol_idx],
-                        line[file_idx] ,header.index("achr38"), header.index("apos38"), header.index("REF") ,header.index("ALT") ,
-                        header.index("beta"), header.index("pval") )
+                        line[file_idx] ,header.index(self.column_names["chr"]), header.index(self.column_names["pos"]), header.index(self.column_names["ref"]) ,header.index(self.column_names["alt"]) ,
```
not a crucial one but could create dict of colname->index to make code shorter. Also does not need to scan through all columns although in single file there aren't many anyway.

Fedja commented on pheweb/serve/data_access/db.py:
```diff
@@ -1450,5 +1472,15 @@ def get_UKBB_dao(self, singlematrix=False):
+        retval = {}
+        retval["externalresultmatrix"] = "externalresultmatrix" in self.dao_impl
+        retval["externalresult"] = "externalresult" in self.dao_impl
+        return retval
```
namedtuple would be better here so the fields are fixed.

Lipastomies replied:
True. Another possibility would be to move the logic from [here](https://github.com/FINNGEN/pheweb/blob/6da6a611168c65d74218807dad32c151e3531012/pheweb/serve/server_jeeves.py#L30-L40) to the datafactory direct so jeeves would become something like
```python
class ServerJeeves(object):
    def __init__(self, server_conf):
        #...
        self.ukbb_dao, self.ukbb_matrixdao = self.dbs_fact.get_UKBB_objects()
        #...
```

and in factory have
```python
class DataFactory(object):
#...
    def get_UKBB_objects(self) -> Tuple[Optional[ExternalResultDb],Optional[ExternalResultDb]]:
        file_available = "externalresult" in self.dao_impl
        matrix_available = "externalresultmatrix" in self.dao_impl
        if file_available and matrix_available:
            return (self.dao_impl["externalresult"], self.dao_impl["externalresultmatrix"])
        elif file_available:
            return (self.dao_impl["externalresult"], self.dao_impl["externalresult"])
         elif matrix_available:
             return (self.dao_impl["externalresultmatrix"], self.dao_impl["externalresultmatrix"])
         else:
             return (None, None)
```

Then we don't have to create a new dictionary or namedtuple to tell which of those DAOs are available, and the decision of what ukbb dao to use would be on DataFactory. It would also be more difficult to accidentally use a wrong DAO.

Fedja commented on config/config.r5.py:
```diff
@@ -36,12 +36,40 @@
         "externalresultmatrix": {
-            "ExternalMatrixResultDao": {"matrix":"/mnt/nfs/ukbb_neale/matrix.tsv.gz", "metadatafile":"/mnt/nfs/ukbb_neale/ukbb_r1_match_pheno_dup_correct_simple_meta.tsv"}
+            "ExternalMatrixResultDao": {"matrix":"/mnt/nfs/ukbb_neale/matrix.tsv.gz", "metadatafile":"/mnt/nfs/ukbb_neale/ukbb_r1_match_pheno_dup_correct_simple_meta.tsv",
+            "column_names":{"pval":"pval","beta":"beta"}}
```
yep why not add the cpra column config here also

Fedja commented:
I think its good enough to runtime figure the chr dickery as you have done.
Also table is pretty enough

Lipastomies commented:
> @Lipastomies add OR also to gene page main table and coding variant table.
> also gene report download seems to error out .... UKBB beta needed there also: http://dev.finngen.fi/gene/CLCN6

Checked the error, seems to be related to latex not being available?
```python
[2021-01-20 07:13:45,458] ERROR in app: Exception on /genereport/CLCN6 [GET]
Traceback (most recent call last):
  File "/home/lehisart/ext_venv/lib/python3.5/site-packages/flask/app.py", line 2447, in wsgi_app
    response = self.full_dispatch_request()
  File "/home/lehisart/ext_venv/lib/python3.5/site-packages/flask/app.py", line 1952, in full_dispatch_request
    rv = self.handle_user_exception(e)
  File "/home/lehisart/ext_venv/lib/python3.5/site-packages/flask/app.py", line 1821, in handle_user_exception
    reraise(exc_type, exc_value, tb)
  File "/home/lehisart/ext_venv/lib/python3.5/site-packages/flask/_compat.py", line 39, in reraise
    raise value
  File "/home/lehisart/ext_venv/lib/python3.5/site-packages/flask/app.py", line 1950, in full_dispatch_request
    rv = self.dispatch_request()
  File "/home/lehisart/ext_venv/lib/python3.5/site-packages/flask/app.py", line 1936, in dispatch_request
    return self.view_functions[rule.endpoint](**req.view_args)
  File "/home/lehisart/pheweb/pheweb/serve/server.py", line 468, in gene_report
    gene_top_assoc_threshold=conf.report_conf["gene_top_assoc_threshold"], func_var_assoc_threshold=conf.report_conf["func_var_assoc_threshold"] )
  File "/home/lehisart/pheweb/pheweb/serve/reporting/report.py", line 43, in render_template
    return build_pdf(tmpl.render(**params))
  File "/home/lehisart/ext_venv/lib/python3.5/site-packages/latex/build.py", line 234, in build_pdf
    raise RuntimeError('No available builder could be instantiated. '
RuntimeError: No available builder could be instantiated. Please make sure LaTeX is installed.
```
Seems like the error is just missing LaTeX on the dev machine. 
We'll have to test this on a docker for sure, but I wouldn'ẗ be too worried about this right now.

Lipastomies commented:
> @Lipastomies add OR also to gene page main table and coding variant table.
> also gene report download seems to error out .... UKBB beta needed there also: http://dev.finngen.fi/gene/CLCN6

It seems UKBB data is not fetched for coding variant table, so I'll have to touch the jeeves function/frontend data grathering a bit. @juhis Do you have any opinions on whether it's better to add a call to ukbb data in the [jeeves coding() function](https://github.com/FINNGEN/pheweb/blob/master/pheweb/serve/server_jeeves.py#L402-L403) or handle it in [frontend](https://github.com/FINNGEN/pheweb/blob/master/pheweb/serve/react/js/components/Coding.js)?

Lipastomies commented:
> > @Lipastomies add OR also to gene page main table and coding variant table.
> > also gene report download seems to error out .... UKBB beta needed there also: http://dev.finngen.fi/gene/CLCN6
> 
> It seems UKBB data is not fetched for coding variant table, so I'll have to touch the jeeves function/frontend data grathering a bit. @juhis Do you have any opinions on whether it's better to add a call to ukbb data in the [jeeves coding() function](https://github.com/FINNGEN/pheweb/blob/master/pheweb/serve/server_jeeves.py#L402-L403) or handle it in [frontend](https://github.com/FINNGEN/pheweb/blob/master/pheweb/serve/react/js/components/Coding.js)?
> 
> If I were to do it in frontend, I'm thinking maybe make a separate function for downloading the relevant UKBB variants and then join these in another function. In any case, I think I'll need to touch frontend columns & download header.

Discussed with Mitja and Mutaamba and we decided to leave the change for later. IF/When we do the change at some point, trhe data wrangling should be in the backend, and frontend should pretty much only render stuff.

Lipastomies commented:
Changed in d58c356:
- Added UKBB OR to latex template & gene report page
- Changed ukbb_dao and ukbb_matrix_dao init to be inside DataFactory, as discussed above
- Added cpra to config for matrix dao

Fedja commented:
I believe all of this will be happening in the new import Mutaamba is finishing so I believe we can delete this PR?
@Lipastomies @majorseitan
